controllers/helmrelease_controller.go

```diff
@@ -18,36 +18,376 @@ package controllers
+		return ctrl.Result{}, err
+	}
+
+	// Check source readiness
+	if source.GetArtifact() == nil {
+		msg := "Source is not ready"
+		hr = v2.HelmReleaseNotReady(hr, hr.Status.LastAttemptedRevision, hr.Status.LastReleaseRevision, v2.ArtifactFailedReason, msg)
+		if err := r.Status().Update(ctx, &hr); err != nil {
+			log.Error(err, "unable to update HelmRelease status")
+			return ctrl.Result{Requeue: true}, err
+		}
+		log.Info(msg)
+		return ctrl.Result{}, nil
+	}
+
+	// Check dependencies
+	if len(hr.Spec.DependsOn) > 0 {
+		if err := r.checkDependencies(hr); err != nil {
+			hr = v2.HelmReleaseNotReady(hr, hr.Status.LastAttemptedRevision, hr.Status.LastReleaseRevision, v2.DependencyNotReadyReason, err.Error())
+			if err := r.Status().Update(ctx, &hr); err != nil {
+				log.Error(err, "unable to update HelmRelease status")
+				return ctrl.Result{Requeue: true}, err
+			}
+			// Exponential backoff would cause execution to be prolonged too much,
+			// instead we requeue on a fixed interval.
+			msg := fmt.Sprintf("Dependencies do not meet ready condition (%s), retrying in %s", err.Error(), r.requeueDependency.String())
+			log.Info(msg)
+			return ctrl.Result{RequeueAfter: r.requeueDependency}, nil
+		}
+		log.Info("All dependencies area ready, proceeding with release")
+	}
 
-	// your logic here
+	reconciledHr, err := r.release(log, hr, source)
+	if err != nil {
+		log.Error(err, "HelmRelease reconciliation failed", "revision", source.GetArtifact().Revision)
+	}
 
-	return ctrl.Result{}, nil
+	if err := r.Status().Update(ctx, &reconciledHr); err != nil {
+		log.Error(err, "unable to update HelmRelease status after reconciliation")
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	// Log reconciliation duration
+	log.Info(fmt.Sprintf("HelmRelease reconcilation finished in %s, next run in %s",
+		time.Now().Sub(start).String(),
+		hr.Spec.Interval.Duration.String(),
+	))
+
+	return ctrl.Result{RequeueAfter: hr.Spec.Interval.Duration}, nil
+}
+
+type HelmReleaseReconcilerOptions struct {
+	MaxConcurrentReconciles   int
+	DependencyRequeueInterval time.Duration
 }
 
-func (r *HelmReleaseReconciler) SetupWithManager(mgr ctrl.Manager) error {
+func (r *HelmReleaseReconciler) SetupWithManager(mgr ctrl.Manager, opts HelmReleaseReconcilerOptions) error {
+	r.requeueDependency = opts.DependencyRequeueInterval
 	return ctrl.NewControllerManagedBy(mgr).
-		For(&helmv2alpha1.HelmRelease{}).
+		For(&v2.HelmRelease{}).
+		WithEventFilter(HelmReleaseReconcileAtPredicate{}).
+		WithEventFilter(HelmReleaseGarbageCollectPredicate{Config: r.Config, Log: r.Log}).
+		WithOptions(controller.Options{MaxConcurrentReconciles: opts.MaxConcurrentReconciles}).
 		Complete(r)
 }
+
+func (r *HelmReleaseReconciler) release(log logr.Logger, hr v2.HelmRelease, source sourcev1.Source) (v2.HelmRelease, error) {
+	// Acquire lock
+	unlock, err := lock(fmt.Sprintf("%s-%s", hr.GetName(), hr.GetNamespace()))
+	if err != nil {
+		err = fmt.Errorf("tmp dir error: %w", err)
```

Comment:
```suggestion
		err = fmt.Errorf("lockfile error: %w", err)
```
I need to change this in kustomize controller as well.

Reply:
It is quite likely that this same wrong message exists in the kustomize-controller.